### PR TITLE
Add opentelemetry support

### PR DESCRIPTION
**Pull request by NikitaSkrynnik**
## Issue link

https://github.com/networkservicemesh/sdk/issues/584

## How Has This Been Tested?

- [ ] Added unit testing to cover
- [ ] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [ ] Bug fix
- [ ] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/chains/nsmgr/single_test.go [hunk 1]
```diff
@@ -37,7 +37,10 @@ import (
 )
 
 func Test_DNSUsecase(t *testing.T) {
-	t.Cleanup(func() { goleak.VerifyNone(t) })
+	t.Cleanup(func() {
+		time.Sleep(time.Second * 20)
+		goleak.VerifyNone(t)
+	})
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgr/single_test.go:43]: Please revert this change

File: pkg/networkservice/common/updatepath/server.go [hunk 2]
```diff
@@ -52,6 +56,19 @@ func (i *updatePathServer) Request(ctx context.Context, request *networkservice.
...
+		var expireTime time.Time
+		tok, expireTime, err = token.FromContext(ctx)
+
+		if err != nil {
+			log.FromContext(ctx).Warnf("an error during getting token from the context: %+v", err)
+		} else {
+			prev.Expires = timestamppb.New(expireTime.Local())
+			prev.Token = tok
+		}
+	}
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/updatepath/server.go:70]: Please revert this change.

File: pkg/tools/log/spanlogger/span.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+// Opentelemetry span
+type otelSpan struct {
+	span opentelemetry.Span
+}
+
+func (otelsp *otelSpan) Log(level, format string, v ...interface{}) {
+	otelsp.span.AddEvent(
+		"",
```
Comment (denis-tingaikin) [inline pkg/tools/log/spanlogger/span.go:89]: Why is this empty?

File: pkg/tools/log/spanlogger/span.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+		opentelemetry.WithAttributes([]attribute.KeyValue{
+			attribute.String("event", level),
+			attribute.String("message", fmt.Sprintf(format, v...)),
+		}...),
+	)
+}
+
+func (otelsp *otelSpan) LogObject(k, v interface{}) {
+	otelsp.span.AddEvent(
+		"",
```
Comment (denis-tingaikin) [inline pkg/tools/log/spanlogger/span.go:99]: Why is this empty?

File: pkg/tools/opentelemetry/README.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
+# OpenTelemetry Collector Example
```
Comment (denis-tingaikin) [inline pkg/tools/opentelemetry/README.md:1]: Nice

File: pkg/tools/opentelemetry/meterhelper/meter_helper.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		}
+		_, ok := m.recorderMap[metricName]
+		if !ok {
+			m.recorderMap[metricName] = metric.Must(m.meter).NewInt64Histogram(
+				m.prefix + metricName,
+			)
+		}
+		m.recorderMap[metricName].Record(ctx, recVal, m.connLabel)
+	}
+}
```
Comment (denis-tingaikin) [inline pkg/tools/opentelemetry/meterhelper/meter_helper.go:74]:
```suggestion
```

File: pkg/tools/opentelemetry/meterhelper/meter_helper.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+			continue
+		}
+		_, ok := m.recorderMap[metricName]
+		if !ok {
+			m.recorderMap[metricName] = metric.Must(m.meter).NewInt64Histogram(
+				m.prefix + metricName,
+			)
+		}
+		m.recorderMap[metricName].Record(ctx, recVal, m.connLabel)
+	}
```
Comment (denis-tingaikin) [inline pkg/tools/opentelemetry/meterhelper/meter_helper.go:73]: Let's use this inline in metrics/server.go

File: pkg/tools/opentelemetry/opentelemetry.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+)
+
+const (
+	// InstrumentationName - denotes the library that provides the instrumentation
+	InstrumentationName = "NSM"
```
Comment (denis-tingaikin) [inline pkg/tools/opentelemetry/opentelemetry.go:40]: Should this be configurable?

File: pkg/networkservice/common/refresh/client.go [hunk 9]
```diff
@@ -59,12 +57,7 @@ func (t *refreshClient) Request(ctx context.Context, request *networkservice.Net
 	}
 
 	// Compute refreshAfter
-	refreshAfter, err := after(ctx, conn)
-	if err != nil {
-		// If we can't refresh, we should close down chain
-		_, _ = t.Close(ctx, conn)
-		return nil, err
-	}
+	refreshAfter := after(ctx, conn)
```
Comment (glazychev-art) [inline pkg/networkservice/common/refresh/client.go:60]: Please revert changes in this file

Comment (glazychev-art) [reply pkg/networkservice/common/refresh/client.go]: I was wrong. Please, resolve this comment

**Review by denis-tingaikin — APPROVED** [review]
In general looks good.

@glazychev-art Could you also have a look?

File: pkg/tools/log/logger.go [hunk 10]
```diff
@@ -19,15 +19,41 @@ package log
...
+	telemetryOT   = "opentracing"
+	telemetryOTel = "opentelemetry"
+
+	// Opentracing enabled by default
+	telemetryDefault = telemetryOTel
 )
 
+// IsOpentracingEnabled returns true if opentracing enabled
+func IsOpentracingEnabled() bool {
+	return telemetryOT == getTelemetryEnv()
```
Comment (glazychev-art) [inline pkg/tools/log/logger.go:41]: Why do we need opentracing now?

**Review by denis-tingaikin — APPROVED** [review]

**Review by denis-tingaikin — APPROVED** [review]
I think this is ready to merge.

@glazychev-art Let us know if you still have any comments on these changes.

**Review by glazychev-art — APPROVED** [review]
